### PR TITLE
Introduces TypeScript

### PR DESCRIPTION
**Pull request by KaiVolland**
This intrdouces TypeScript to the base-util.

- `babel` is removed as the transpilation is done by TS now.
- `documentation` is replaced with `typedoc`
- Basic typings for Utils and Tests are included and exported

### CODE REVIEW

File: src/UrlUtil/UrlUtil.spec.ts [hunk 1]
```diff
@@ -230,8 +231,9 @@ describe('UrlUtil', () => {
           'http://foo@example.com/foo?humpty=dumpty': true,
           'http://foo:s3cretP4ssw0rd@example.com/foo?humpty=dumpty': true
         };
-
+        // @ts-ignore
         for (const [key, value] of Object.entries(urls)) {
+          console.log(key, UrlUtil.isValid(key));
```
Comment (dnlkoch) [inline src/UrlUtil/UrlUtil.spec.ts]: Please remove the logger.

File: src/ObjectUtil/ObjectUtil.spec.ts [hunk 2]
```diff
@@ -103,6 +103,7 @@ describe('ObjectUtil', () => {
         obj[Object.prototype.toString.call(unexpectedKey)] = true;
       });
       unexpectedKeys.forEach(unexpectedKey => {
+        // @ts-ignore
         expect(ObjectUtil.getValue(unexpectedKey, obj)).toBeUndefined();
```
Comment (dnlkoch) [inline src/ObjectUtil/ObjectUtil.spec.ts]: Why do we have to `ts-ignore` this line?

File: package.json [hunk 3]
```diff
@@ -5,23 +5,23 @@
...
-    "start:docs": "documentation serve -w -g src/index.js",
-    "build:docs": "npm run clean:docs && documentation build -g -f html -o build/docs src/index.js",
-    "build:docs-md": "documentation readme src/index.js --section=\"API Documentation\"",
-    "build:dist": "npm run clean:dist && BABEL_ENV=production babel src --out-dir dist --copy-files --ignore spec.js",
+    "build": "npm run test:ci && npm run build:dist && npm run build:docs",
+    "build:docs": "npm run clean:docs && typedoc --out ./doc/ --exclude **/*.spec.**",
+    "build:dist": "tsc -p ./tsconfig.prod.json",
     "coveralls": "cat coverage/lcov.info | coveralls",
     "prepublishOnly": "npm run build:dist",
     "deploy": "NODE_DEBUG=gh-pages node tasks/update-gh-pages.js",
```
Comment (dnlkoch) [inline package.json]: I assume it should not be a problem, but have you tried if the task is still working with the new documentation framework?

Comment (KaiVolland) [reply package.json]: I change the outputdir to `./build/docs` so it should fit the ghpages task now.

File: src/ObjectUtil/ObjectUtil.spec.ts (hunk 2 above)
Comment (KaiVolland) [reply src/ObjectUtil/ObjectUtil.spec.ts]: It wasn't. I removed it.

File: src/UrlUtil/UrlUtil.spec.ts (hunk 1 above)
Comment (KaiVolland) [reply src/UrlUtil/UrlUtil.spec.ts]: done

**Pull request merged**